src/sqlfluff/core/parser/grammar/base.py

```diff
@@ -632,6 +632,84 @@ def __repr__(self):
             ),
         )
 
+    def __eq__(self, other):
+        """Two grammars are equal if their elements and types are equal.
+
+        NOTE: This could potentially mean that two grammars with
+        the same elements but _different configuration_ will be
+        classed as the same. If this matters for your use case,
+        consider extending this function.
+        """
+        return type(self) is type(other) and self._elements == other._elements
+
+    def copy(
+        self,
+        insert: Optional[list] = None,
+        at: Optional[int] = None,
+        before: Optional[Any] = None,
+        remove: Optional[list] = None,
+    ):
+        """Create a copy of this grammar, optionally with differences.
+
+        This is mainly used in dialect inheritance.
+
+
+        Args:
+            insert (:obj:`list`, optional): Matchable elements to
+                insert. This is inserted pre-expansion so can include
+                unexpanded elements as normal.
+            at (:obj:`int`, optional): The position in the elements
+                to insert the item. Defaults to `None` which means
+                insert at the end of the elements.
+            before (optional): An alternative to _at_ to determine the
```

Comment:
Why is this `Optional[Any]` rather than `Optional[int]` like `at`? I have an alternate idea that _may_ be clearer... Could we eliminate the `before` option and pass a _negative number_ to `at` for this purpose?

Reply:
On the datatype - not sure, I think that might just be an oversight. 🤔 

On the `before` option - at the moment you can use `after` with a negative number and it will insert it after _an index from the end_. I might see if I can do something similar to what you suggest though...

Reply:
Ah - no - this *is* intentional!

`before` allows you to provide a matching element to look for, in case you don't know it's index. I'll provide an example to show that.